router/proxy/reverseproxy.go

```diff
@@ -0,0 +1,332 @@
+	transport := p.transport
+	if transport == nil {
+		panic("router: nil transport for proxy")
+	}
+
+	res, uconn, err := transport.UpgradeHTTP(req)
+	if err != nil {
+		p.logf("router: proxy error: %v", err)
+		rw.WriteHeader(http.StatusServiceUnavailable)
+		rw.Write(serviceUnavailable)
+		return
+	}
+	defer uconn.Close()
+
+	p.writeResponse(rw, res)
+	res.Body.Close()
+
+	if res.StatusCode != 101 {
+		return
+	}
+
+	hj, ok := rw.(http.Hijacker)
+	if !ok {
+		p.logf("router: hijack failed")
+		return
+	}
+	dconn, bufrw, err := hj.Hijack()
+	if err != nil {
+		p.logf("router: hijack failed: %v", err)
+		return
+	}
+	joinConns(uconn, &streamConn{bufrw.Reader, dconn})
+}
+
+func (p *ReverseProxy) writeResponse(rw http.ResponseWriter, res *http.Response) {
+	// remove hop-by-hop headers as specified in connection options.
+	for _, h := range hopHeaders {
+		res.Header.Del(h)
+	}
+
+	// remove the Upgrade header and headers referenced in the Connection
+	// header if HTTP < 1.1 or if Connection header didn't contain "upgrade":
+	// https://tools.ietf.org/html/rfc7230#section-6.7
+	if !res.ProtoAtLeast(1, 1) || !isConnectionUpgrade(res.Header) {
+		res.Header.Del("Upgrade")
+
+		// A proxy or gateway MUST parse a received Connection header field before a
+		// message is forwarded and, for each connection-option in this field, remove
+		// any header field(s) from the message with the same name as the
+		// connection-option, and then remove the Connection header field itself (or
+		// replace it with the intermediary's own connection options for the
+		// forwarded message): https://tools.ietf.org/html/rfc7230#section-6.1
+		tokens := strings.Split(res.Header.Get("Connection"), ",")
+		for _, hdr := range tokens {
+			res.Header.Del(hdr)
+		}
+		res.Header.Del("Connection")
+	}
+
+	copyHeader(rw.Header(), res.Header)
+
+	rw.WriteHeader(res.StatusCode)
+	p.copyResponse(rw, res.Body)
+}
+
+func isConnectionUpgrade(h http.Header) bool {
+	for _, token := range strings.Split(h.Get("Connection"), ",") {
+		if v := strings.ToLower(strings.TrimSpace(token)); v == "upgrade" {
+			return true
+		}
+	}
+	return false
+}
+
+func (p *ReverseProxy) copyResponse(dst io.Writer, src io.Reader) {
+	if p.FlushInterval != 0 {
+		if wf, ok := dst.(writeFlusher); ok {
+			mlw := &maxLatencyWriter{
+				dst:     wf,
+				latency: p.FlushInterval,
+				done:    make(chan bool),
+			}
+			go mlw.flushLoop()
+			defer mlw.stop()
+			dst = mlw
+		}
+	}
+
+	io.Copy(dst, src)
+}
+
+func (p *ReverseProxy) logf(format string, args ...interface{}) {
+	if p.ErrorLog != nil {
+		p.ErrorLog.Printf(format, args...)
+	} else {
+		log.Printf(format, args...)
+	}
+}
+
+func copyHeader(dst, src http.Header) {
+	for k, vv := range src {
+		for _, v := range vv {
+			dst.Add(k, v)
+		}
+	}
+}
+
+type closeWriter interface {
+	CloseWrite() error
+}
+
+func closeWrite(conn net.Conn) {
+	if cw, ok := conn.(closeWriter); ok {
+		cw.CloseWrite()
+	} else {
+		conn.Close()
+	}
```

Comment:
yes, `streamConn` is no longer a `closeWriter`. Previously, it had a shim `CloseWrite()` that just called `Close()`.